rust/xaynet-server/src/services/messages/multipart.rs

```diff
@@ -0,0 +1,456 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    task::Poll,
+};
+
+use futures::{
+    future::{self, Ready},
+    task::Context,
+};
+use tower::Service;
+use xaynet_core::{
+    crypto::{PublicEncryptKey, PublicSigningKey},
+    message::{Chunk, DecodeError, FromBytes, Message, Payload, Sum, Sum2, Tag, Update},
+};
+
+use crate::services::messages::ServiceError;
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone))]
+pub struct MultipartMessage {
+    /// Public key of the participant sending the message
+    participant_pk: PublicSigningKey,
+    /// Public key of the coordinator
+    coordinator_pk: PublicEncryptKey,
+    /// Message type
+    tag: Tag,
+    /// The ID of the last chunk is actually the total number of
+    /// chunks this message is made of.
+    last_chunk_id: Option<u16>,
+    /// Chunks, ordered by ID
+    data: BTreeMap<u16, Vec<u8>>,
+}
+
+impl MultipartMessage {
+    fn new(tag: Tag, participant_pk: PublicSigningKey, coordinator_pk: PublicEncryptKey) -> Self {
+        MultipartMessage {
+            tag,
+            participant_pk,
+            coordinator_pk,
+            data: BTreeMap::new(),
+            last_chunk_id: None,
+        }
+    }
+    /// Return `true` if the message is complete, _i.e._ if all
+    /// `MultiparMessage` holds all the chunks.
+    fn has_all_chunks(&self) -> bool {
+        self.last_chunk_id
+            .map(|expected_number_of_chunks| self.data.len() == expected_number_of_chunks as usize)
+            .unwrap_or(false)
+    }
+
+    /// Return the first missing chunk. If no chunk is missing, return `None`.
+    #[allow(dead_code)]
+    fn first_missing_chunk(&self) -> Option<u16> {
+        if self
+            .last_chunk_id
+            .map(|last_id| self.data.len() >= last_id as usize)
+            .unwrap_or(false)
+        {
+            return None;
+        }
+
+        let mut current_id = 0;
+        for id in self.data.keys() {
+            if id - current_id > 1 {
+                break;
+            }
+            current_id = *id;
+        }
+        Some(current_id + 1)
+    }
+
+    fn add_chunk(&mut self, chunk: Chunk) {
+        let Chunk { id, last, data, .. } = chunk;
+        if last {
+            self.last_chunk_id = Some(id);
+        }
+        self.data.insert(id, data);
+    }
+
+    /// Aggregate all the chunks. This method should only be called
+    /// when all the chunks are here, otherwise the aggregated message
+    /// will be invalid.
+    fn into_message(self) -> Result<Message, DecodeError> {
+        // FIXME: this is highly inefficient. We end up allocating a
```

Comment:
@finiteprods @Robert-Steiner this is the issue I mentioned. I think the fix is not trivial. We'd have to rework how messages are parsed.